### PR TITLE
Support cloud sql private ip

### PR DESCRIPTION
**Pull request by yhuang**
Here is my first attempt at adding Cloud SQL private network support.  I am new to Go, so I will need some guidance on how to improve this code and add the necessary tests.  Also, I could not disable public IP even when I set ip_configuration.ipv4_enabled to false.

### CODE REVIEW

File: google-beta/resource_sql_database_instance.go [hunk 1]
```diff
@@ -614,10 +626,13 @@ func expandIpConfiguration(configured []interface{}) *sqladmin.IpConfiguration {
 	}
 
 	_ipConfiguration := configured[0].(map[string]interface{})
+
 	return &sqladmin.IpConfiguration{
 		Ipv4Enabled:        _ipConfiguration["ipv4_enabled"].(bool),
 		RequireSsl:         _ipConfiguration["require_ssl"].(bool),
+		PrivateNetwork:     _ipConfiguration["private_network"].(string),
 		AuthorizedNetworks: expandAuthorizedNetworks(_ipConfiguration["authorized_networks"].(*schema.Set).List()),
+		ForceSendFields:    []string{"Ipv4Enabled"},
```
Comment (yhuang) [inline google-beta/resource_sql_database_instance.go]: Adding the `ForceSendFields` in this return statement seemed to have done the trick.  Now instance create and update will enable and disable public IP as expected.

**Comment by yhuang** [comment]
@rileykarson @chrisst @danawillow : I see the comment on this PR:

https://github.com/terraform-providers/terraform-provider-google-beta/pull/21

I do not quite understand how modular magician fits into the flow of PRs against the beta providers.  If I need to open a PR against the upstream file [here](https://github.com/terraform-providers/terraform-provider-google-beta/blob/181c7ede8752c2da8368d8d44d398545e846eb26/google-beta/resource_sql_database_instance.go), I will be happy to do so.  I have tested my changes locally, and it works.  I just need some guidance on where to submit the PR and then help to make sure I have sufficient test coverage.  Thank you.

**Comment by chrisst** [comment]
Hey, we are still in the process of figuring out how to manage the split of our providers so apologies but I'll start with state we're in right now. Our eventual goal is to be generating all of our terraform provider code programmatically and several of our resources are already done this way (you can check the file headers to see which). To get there and also to be able to generate the beta & ga providers from the same code base, we've moved the source of truth of all terraform code to the [magic modules](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/provider/terraform/resources) repository. The magic modules codebase now generates the code for both the google beta and the google provider, so it can effectively be considered the upstream.

So to answer your question, you can create a pull request against https://github.com/terraform-providers/terraform-provider-google (or https://github.com/terraform-providers/terraform-provider-google-beta if it's using beta features). We will continue reviewing code in those repos for resources that aren't completely auto generated, but it will eventually need to be upstreamed to magic modules and we're still figuring out how that process will go. You are also welcome to open it directly against the upstream files in the magic modules repo and we welcome contributions to that project as well.

**Comment by craigatgoogle** [comment]
@yhuang thanks for doing this, this saves us some time. I'm going to submit a PR soon which provides coverage for the VPC Peering Connection resource, this will enable you to implement a proper acctest for this feature in this PR. I'll post an update once my PR is ready.

**Comment by yhuang** [comment]
Thank you, @craigatgoogle.  I cannot wait to rip this code out of my other repo that is using cloud sql private IP currently, once your PR is ready:

```
resource "null_resource" "vpc_to_services_peering" {
  provisioner "local-exec" {
    command = <<EOF
gcloud beta services vpc-peerings connect \
--service=servicenetworking.googleapis.com \
--network=${local.network_name} \
--ranges=${google_compute_global_address.managed_services.name} \
--project=${local.project_id};
EOF
  }

  provisioner "local-exec" {
    when = "destroy"

    command = <<EOF
gcloud compute networks peerings delete servicenetworking-googleapis-com \
--network=${local.network_name} \
--project=${local.project_id}
EOF
  }
}```

**Comment by craigatgoogle** [comment]
Here we go: #84

File: google-beta/resource_sql_database_instance.go [hunk 2]
```diff
@@ -256,7 +264,7 @@ func resourceSqlDatabaseInstance() *schema.Resource {
 				ForceNew: true,
 			},
 
-			"ip_address": &schema.Schema{
+			"ip_addresses": &schema.Schema{
```
[the remark this answers is not in the document]
Comment (yhuang) [reply google-beta/resource_sql_database_instance.go]: @danawillow : I was going by what was in the documentation:

https://cloud.google.com/sql/docs/mysql/admin-api/v1beta4/instances#resource

```
{
  "kind": "sql#instance",
  "selfLink": string,
  "name": string,
  "connectionName": string,
  "etag": string,
  "project": string,
  "state": string,
  "backendType": string,
  "databaseVersion": string,
  "region": string,
  "currentDiskSize": long,
  "maxDiskSize": long,
  "settings": {
    "kind": "sql#settings",
    "settingsVersion": long,
    "authorizedGaeApplications": [
      string
    ],
    "tier": string,
    "backupConfiguration": {
      "kind": "sql#backupConfiguration",
      "startTime": string,
      "enabled": boolean,
      "binaryLogEnabled": boolean,
      "replicationLogArchivingEnabled": boolean
    },
    "pricingPlan": string,
    "replicationType": string,
    "activationPolicy": string,
    "ipConfiguration": {
      "ipv4Enabled": boolean,
      "authorizedNetworks": [
        {
          "kind": "sql#aclEntry",
          "value": string,
          "name": string,
          "expirationTime": datetime
        }
      ],
      "requireSsl": boolean,
      "privateNetwork": string
    },
    "locationPreference": {
      "kind": "sql#locationPreference",
      "followGaeApplication": string,
      "zone": string
    },
    "databaseFlags": [
      {
        "name": string,
        "value": string
      }
    ],
    "databaseReplicationEnabled": boolean,
    "crashSafeReplicationEnabled": boolean,
    "dataDiskSizeGb": long,
    "dataDiskType": string,
    "maintenanceWindow": {
      "kind": "sql#maintenanceWindow",
      "hour": integer,
      "day": integer,
      "updateTrack": string
    },
    "storageAutoResize": boolean,
    "storageAutoResizeLimit": long,
    "availabilityType": string,
    "userLabels": {
      (key): string
    }
  },
  "serverCaCert": sslCerts Resource,
  "ipAddresses": [
    {
      "ipAddress": string,
      "timeToRetire": datetime,
      "type": string
    }
  ],
  "instanceType": string,
  "masterInstanceName": string,
  "replicaNames": [
    string
  ],
  "failoverReplica": {
    "name": string,
    "available": boolean
  },
  "ipv6Address": string,
  "serviceAccountEmailAddress": string,
  "onPremisesConfiguration": {
    "kind": "sql#onPremisesConfiguration",
    "hostPort": string
  },
  "replicaConfiguration": {
    "kind": "sql#replicaConfiguration",
    "mysqlReplicaConfiguration": {
      "kind": "sql#mysqlReplicaConfiguration",
      "dumpFilePath": string,
      "username": string,
      "password": string,
      "connectRetryInterval": integer,
      "masterHeartbeatPeriod": long,
      "caCertificate": string,
      "clientCertificate": string,
      "clientKey": string,
      "sslCipher": string,
      "verifyServerCertificate": boolean
    },
    "failoverTarget": boolean
  },
  "suspensionReason": [
    string
  ],
  "gceZone": string
}
```

[the remark this answers is not in the document]
Comment (yhuang) [reply google-beta/resource_sql_database_instance.go]: Thank you @danawillow for explaining in real-time that Terraform resource parameters do not need to be the same as their corresponding fields in the API schema.

**Comment by rileykarson** [comment]
Closing this out as I believe we've got support now.

**Pull request closed without merging**